presto-main/src/main/java/com/facebook/presto/sql/gen/TryCodeGenerator.java

```diff
@@ -0,0 +1,140 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.invoke.MethodType.methodType;
+
+public class TryCodeGenerator
+        implements BytecodeGenerator
+{
+    private static final String EXCEPTION_HANDLER_NAME = "tryExpressionExceptionHandler";
+
+    private final Map<CallExpression, MethodDefinition> tryMethodsMap;
+    private final List<? extends Variable> inputParameters;
+
+    private static int tryMethodId = 1;
+
+    public TryCodeGenerator(Map<CallExpression, MethodDefinition> tryMethodsMap, List<? extends Variable> inputParameters)
+    {
+        this.tryMethodsMap = tryMethodsMap;
+        this.inputParameters = inputParameters;
+    }
+
+    @Override
+    public BytecodeNode generateExpression(Signature signature, BytecodeGeneratorContext context, Type returnType, List<RowExpression> arguments)
+    {
+        checkArgument(arguments.size() == 1, "try methods only contain a single expression");
+        checkArgument(getOnlyElement(arguments) instanceof CallExpression, "try methods must contain a call expression");
+
+        CallExpression innerCallExpression = (CallExpression) getOnlyElement(arguments);
+        checkState(tryMethodsMap.containsKey(innerCallExpression), "try methods map does not contain this try call");
+
+        BytecodeBlock bytecodeBlock = new BytecodeBlock()
+                .comment("load required variables")
+                .getVariable(context.getScope().getVariable("this"));
+
+        inputParameters.stream()
+                .forEach(bytecodeBlock::getVariable);
+
+        bytecodeBlock.comment("call dynamic try method: " + tryMethodsMap.get(innerCallExpression).getName())
+                .invokeVirtual(tryMethodsMap.get(innerCallExpression))
+                .append(unboxPrimitiveIfNecessary(context.getScope(), Primitives.wrap(innerCallExpression.getType().getJavaType())));
+
+        return bytecodeBlock;
+    }
+
+    public static MethodDefinition defineTryMethod(
+            ClassDefinition classDefinition,
+            BytecodeExpressionVisitor innerExpressionVisitor,
+            RowExpression innerRowExpression,
+            List<Parameter> inputParameters,
+            Class<?> returnType,
+            CallSiteBinder callSiteBinder,
+            CachedInstanceBinder cachedInstanceBinder,
+            FunctionRegistry functionRegistry)
+    {
+        MethodDefinition method = classDefinition.declareMethod(a(PUBLIC), "try_" + (tryMethodId++), type(returnType), inputParameters);
+        Scope calleeMethodScope = method.getScope();
+
+        BytecodeGeneratorContext calleeMethodContext = new BytecodeGeneratorContext(innerExpressionVisitor, calleeMethodScope, callSiteBinder, cachedInstanceBinder, functionRegistry);
+        BytecodeNode innerExpression = calleeMethodContext.generate(innerRowExpression);
```

Comment:
It doesn't look like you need the generator context here if you are only using the `generate` method. That should reduce a bunch of arguments to `defineTryMethod` function

        innerRowExpression.accept(innerExpressionVisitor, calleeMethodScope);

This works because the additional parameters you are passing to the constructor `BytecodeGeneratorContext` had been previously passed to `BytecodeExpressionVisitor`.